data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsService.java

```diff
@@ -0,0 +1,221 @@
+    public static final String REQUESTS_SUCCEEDED = "cloudWatchLogsRequestsSucceeded";
+    public static final String REQUESTS_FAILED = "cloudWatchLogsRequestsFailed";
+    private static final Logger LOG = LoggerFactory.getLogger(CloudWatchLogsService.class);
+    private final CloudWatchLogsClient cloudWatchLogsClient;
+    private final Buffer buffer;
+    private final ThresholdCheck thresholdCheck;
+    private final List<EventHandle> bufferedEventHandles;
+    private final String logGroup;
+    private final String logStream;
+    private final int retryCount;
+    private final long backOffTimeBase;
+    private final io.micrometer.core.instrument.Counter logEventSuccessCounter; //Counter to be used on the fly for counting successful transmissions. (Success per single event successfully published).
+    private final Counter requestSuccessCount;
+    private final io.micrometer.core.instrument.Counter logEventFailCounter;
+    private final io.micrometer.core.instrument.Counter requestFailCount; //Counter to be used on the fly during error handling.
+    private int failCounter = 0;
+    private boolean failedPost;
+    private final StopWatch stopWatch;
+    private boolean stopWatchOn;
+    private final ReentrantLock reentrantLock;
+
+    public CloudWatchLogsService(final CloudWatchLogsClient cloudWatchLogsClient, final CloudWatchLogsSinkConfig cloudWatchLogsSinkConfig, final Buffer buffer,
+                                 final PluginMetrics pluginMetrics, final ThresholdCheck thresholdCheck, final int retryCount, final long backOffTimeBase) {
+
+        this.cloudWatchLogsClient = cloudWatchLogsClient;
+        this.buffer = buffer;
+        this.logGroup = cloudWatchLogsSinkConfig.getLogGroup();
+        this.logStream = cloudWatchLogsSinkConfig.getLogStream();
+        this.thresholdCheck = thresholdCheck;
+
+        this.retryCount = retryCount;
+        this.backOffTimeBase = backOffTimeBase;
+
+        this.bufferedEventHandles = new ArrayList<>();
+        this.logEventSuccessCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_PUSHED_TO_CWL_SUCCESS);
+        this.requestFailCount = pluginMetrics.counter(REQUESTS_FAILED);
+        this.logEventFailCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_PUSHED_TO_CWL_FAIL);
+        this.requestSuccessCount = pluginMetrics.counter(REQUESTS_SUCCEEDED);
+
+        reentrantLock = new ReentrantLock();
+
+        stopWatch = StopWatch.create();
+        stopWatchOn = false;
+    }
+
+    /**
+     * Function handles the packaging of events into log events before sending a bulk request to CloudWatchLogs.
+     * Implements simple conditional buffer. (Sends once batch size, request size in bytes, or time limit is reached)
+     * @param logs - Collection of Record events which hold log data.
+     */
+    public void output(final Collection<Record<Event>> logs) {
+        reentrantLock.lock();
```

Comment:
A user can define multiple worker threads, and there is a sink thread per worker thread. Locking these will work fine, but it may impact the performance. Can the lock's scope be reduced at all?

Reply:
I am a bit unclear on how the actual pipeline instance works. But if there is only one sink thread per worker, is there no risk of a race condition? If so can we just avoid the lock.

Reply:
The user can configure some number of pipeline workers. This is the total number of threads running. Thus, you may have N threads calling `output` on this instance. So there is a need for thread synchronization.

I'm wondering if we can reduce the scope of the lock so that only part of the work needed here is locked.

One possible solution: Synchronize the calls to write to the buffer. Then have a blocking queue for the log pushing part. In this way, some threads could write to the new buffer while one thread is responsible for pushing to CloudWatch Logs.

This is something we can improve upon later though.